lib/actions/targetActions.js

```diff
@@ -101,15 +106,68 @@ export function dfuImagesUpdateAction(dfuImages) {
     };
 }
 
-export function writeProgressFinishedAction() {
+export function writingStartAction() {
+    return {
+        type: WRITING_START,
+    };
+}
+
+export function writingEndAction() {
+    return {
+        type: WRITING_END,
+    };
+}
+
+export function erasingStartAction() {
+    return {
+        type: ERASING_START,
+    };
+}
+
+export function erasingEndAction() {
+    return {
+        type: ERASING_END,
+    };
+}
+
+export function loadingStartAction() {
     return {
-        type: WRITE_PROGRESS_FINISHED,
+        type: LOADING_START,
     };
 }
 
-export function writeProgressStartAction() {
+export function loadingEndAction() {
     return {
-        type: WRITE_PROGRESS_START,
+        type: LOADING_END,
+    };
+}
+
+export function openDevice(device) {
+    return dispatch => {
+        dispatch(loadingStartAction());
+
+        const { serialNumber, serialport } = device;
+
+        dispatch(targetPortChangedAction(serialNumber, serialport.comName));
+
+        if (device.traits.includes('jlink')) {
+            dispatch(jlinkTargetActions.loadDeviceInfo(serialNumber));
+            return;
+        }
+        if (device.traits.includes('nordicUsb')) {
+            dispatch(usbsdfuTargetActions.openDevice(device));
+            return;
+        }
+
+        const { vendorId, productId } = serialport;
+        const vid = parseInt(vendorId.toString(16), 16);
+        const pid = parseInt(productId.toString(16), 16);
+        if (vid === VendorId.NORDIC_SEMICONDUCTOR && USBProductIds.includes(pid)) {
+            dispatch(usbsdfuTargetActions.openDevice(device));
+            return;
+        }
+
+        logger.error('Unsupported device. The detected device is neither JLink device nor Nordic USB device.');
```

Comment:
Also note that the app's config should prevent to reach this line, so the user probably never sees this message.

Reply:
I updated the text. Ideally the user should never reach this line. But in the code logic it may reach it, and it appeared once with the JLink Lite if I remember correctly. Let's keep it for sometime and check in practice until we are sure the config is robust enough.

Reply:
It may be unlikely to get to this line, but there have been similar cases previously in other products. Just a minor issue. And the updated version looks good :-)